graphene_plugin.py

```diff
@@ -1,381 +1,690 @@
+
+        self._objecttypes: Dict[str, ObjectTypeInfo] = {}
+
+    def get_base_class_hook(self, fullname: str) -> Optional[Callable[[ClassDefContext], None]]:
+        def collect_objecttype_subclass(ctx: ClassDefContext) -> None:
+            """
+            Collect type information about graphene `ObjectType` child classes. This plugin is invoked
+            during semantic analysis.
+            """
+
+            if not ctx.api.final_iteration:
                 ctx.api.defer()
                 return
 
-            # TODO: Improve performance by only getting type info for attributes that have
-            # names matching resolvers.
-            runtime_type = objecttype_base.args[0]
-            attributes_or_none = [
-                create_attribute_type(ctx, attribute_node)
-                for attribute_node in class_body
-                if isinstance(attribute_node, AssignmentStmt)
-            ] + interface_attributes
-            attributes = [
-                attribute_or_none for attribute_or_none in attributes_or_none if attribute_or_none is not None
-            ]
-
-            resolvers = [ # TODO: Maybe use dicts to increase comparison speed
-                create_resolver_type(ctx, resolver_node)
-                for resolver_node in class_body
-                # The check for Decorator assumes we're using `staticmethod`s.
-                if isinstance(resolver_node, (Decorator, FuncDef)) and resolver_node.name.startswith(RESOLVER_PREFIX)
-            ]
-
-            for resolver in resolvers:
-                matching_attributes = [attribute for attribute in attributes if attribute == resolver]
-                assert resolver.context is not None
-                if not matching_attributes:
-                    ctx.api.fail(f'No field with name "{resolver.name}" defined', resolver.context)
+            module = ctx.api.modules[ctx.cls.info.module_name]
+
+            self._objecttypes[ctx.cls.info.fullname] = ObjectTypeInfo.for_classdef(ctx.api, ctx.cls)
+
+            # Here is the fun hack. We want to type-check our `ObjectType`s at type-checking time, but
+            # the `get_base_class_hook` only runs at semantic analysis time and as of now there is no
+            # classdef hook that runs at type-checking time. To work around this, we do two things:
+            #
+            # * First, we add a dummy attribute to the `ObjectType` child class (__graphene_plugin_noop__)
+            # * Then, we add some statements to the module's AST that effectively do this:
+            #   `cast(MyObjectType, None).__graphene_plugin_noop__`
+            #
+            # Because mypy has a `get_attribute_hook` plugin hook that is invoked during type-checking time
+            # we now can rely on that hook to perform our type-checking of the `ObjectType`.
+            _add_var_to_class(NOOP_ATTR_NAME, NoneType(), ctx.cls.info)
+            _add_attr_access_to_module(module, ctx.cls.info, NOOP_ATTR_NAME)
```

Comment:
WOW, this is clever. So we couldn't go with the function call hook after all?

Reply:
The problem with doing that is it made it look like all the type errors were coming from the one file where the `Schema()` function call was happening. And then, even when I fixed that by manually passing in a `file` when calling `.fail()`, that didn't change the fact that type-checking of all the `ObjectType`s wasn't actually happening unless you type-checked the file where the schema was being created. The result is that doing something like `mypy file/with/my/objecttype.py` could look like it was totally successful. But doing `mypy file/with/my/schema.py` could show errors in `file/with/my/objecttype.py`.

This told me that what I was trying to do (create errors in files other than the one that was actually being type-checked) was very unusual and unidiomatic. So I refocused my efforts on figuring out how I could get mypy to call our plugin during type-checking time for each file with an `ObjectType`. And this is what I got!

Reply:
Gotcha, makes sense. Good thinking to do this workaround, it wouldn't have occurred to me. Let's hopefully never do it again!

Reply:
And let's hopefully remove it soon when mypy adds the hook we're looking for!